pytket/extensions/cutensornet/tnstate/simulation.py

```diff
@@ -0,0 +1,341 @@
+            partition[2 * key] = groupA
+            partition[2 * key + 1] = groupB
+
+        # Stop if all groups have less than ``max_q_per_leaf`` qubits in them
+        stop_bisec = all(len(group) <= max_q_per_leaf for group in partition.values())
+
+    qubit_partition = {key: list(leaf_qubits) for key, leaf_qubits in partition.items()}
+    return qubit_partition
+
+
+def _get_sorted_gates(
+    circuit: Circuit,
+    algorithm: SimulationAlgorithm,
+    qubit_partition: Optional[dict[int, list[Qubit]]] = None,
+) -> list[Command]:
+    """Sorts the list of gates so that there's less canonicalisation during simulation.
+
+    Returns an equivalent list of commands fixing the order of parallel gates so that
+    2-qubit gates that are close together are applied one after the other. This reduces
+    the overhead of canonicalisation during simulation.
+
+    Args:
+        circuit: The original circuit.
+        algorithm: The simulation algorithm that will be used on this circuit.
+        qubit_partition: For TTN simulation algorithms only. A partition of the
+            qubits in the circuit into disjoint groups, describing the hierarchical
+            structure of the TTN.
+
+    Returns:
+        The same gates, ordered in a beneficial way for the given algorithm.
+    """
+    all_gates = circuit.get_commands()
+    sorted_gates = []
+    # Entries from `all_gates` that are not yet in `sorted_gates`
+    remaining = set(range(len(all_gates)))
+
+    # Do some precomputation depending on the algorithm
+    if algorithm in [SimulationAlgorithm.TTNxGate]:
+        if qubit_partition is None:
+            raise RuntimeError("You must provide a qubit partition!")
+
+        leaf_of_qubit: dict[Qubit, int] = dict()
+        for leaf, qubits in qubit_partition.items():
+            for q in qubits:
+                leaf_of_qubit[q] = leaf
+
+    elif algorithm in [SimulationAlgorithm.MPSxGate, SimulationAlgorithm.MPSxMPO]:
+        idx_of_qubit = {q: i for i, q in enumerate(circuit.qubits)}
+
+    else:
+        raise RuntimeError(f"Sorting gates for {algorithm} not supported.")
+
+    # Create the list of indices of gates acting on each qubit
+    gate_indices: dict[Qubit, list[int]] = defaultdict(list)
+    for i, g in enumerate(all_gates):
+        for q in g.qubits:
+            gate_indices[q].append(i)
+    # Schedule all 1-qubit gates at the beginning of the circuit
+    for q, indices in gate_indices.items():
+        while indices and len(all_gates[indices[0]].qubits) == 1:
+            i = indices.pop(0)
+            sorted_gates.append(all_gates[i])
+            remaining.remove(i)
+
+    # Decide which 2-qubit gate to apply next
+    last_qubits = [circuit.qubits[0], circuit.qubits[0]]  # Arbitrary choice at start
+    while remaining:
+        # Gather all gates that have nothing in front of them at one of its qubits
+        reachable_gates = [gates[0] for gates in gate_indices.values() if gates]

```

Comment:
Won't that also select 1-qubit gates?...

Reply:
Ohh or is line 276 'pops' them from `gate_indices` too?... Sorry being stupid...